reps/2023-12-04-accelerated-dag.md

```diff
@@ -0,0 +1,360 @@
+tasks.
+
+To support compiled DAGs, we need to make a few assumptions:
+
+-   The application can be expressed as a handful of
+    [DAGs](https://docs.ray.io/en/latest/ray-core/ray-dag.html).
+
+-   Tasks may dynamically call Ray tasks as usual, but the execution of
+    these nested tasks will not be accelerated.
+
+-   A DAG must be declared before it can be executed. The initial
+    declaration may be higher latency than a single DAG execution.
+
+    -   For now, dynamic control/data flow within a DAG will not be
+        allowed. These could eventually be supported by splitting a
+        DAG into sub-DAGs, and using the driver to coordinate the
+        control/data flow between sub-DAGs.
+
+    -   The max size of task outputs must be declared.
+
+    -   A task output can only be read by the downstream DAG nodes,
+        unless it is one of the final outputs of the DAG. (i.e. one
+        cannot access intermediate outputs through the usual ObjectRef
+        API).
+
+### Fault tolerance
+
+Initially, we will provide failure detection at the DAG level. In
+particular, if any task in the DAG fails due to application exception or
+worker death, we will propagate the error to the DAG outputs.
+
+Currently, Ray Core also provides automatic task-level re-execution. At
+the moment, we do not plan to support this, as it would likely require
+extra overhead from tracking intermediate task outputs. However,
+automatic DAG-level re-execution may be possible.
+
+## Application-defined transports
+
+***Goal:** Give the application greater control over the communication
+method, but allow Ray Core to schedule the communication.* This could
+allow us to keep current Ray Core system optimizations such as:
+
+1.  Overlapping communication with task execution.
+2.  Handle communication failures gracefully through retries.
+3.  Control communication concurrency / resource usage.
+
+Accelerated DAGs will ship with a default transport for objects passed
+along the DAG's edges. This will use shared memory if the
+sender/receiver are colocated, and shared memory/Ray Core for
+cross-node.
+
+To support specialized transports, we will extend the DAG API to allow
+application-defined transports that can be called from either Python or
+C++. The exact API is TBD, but here is a simple initial proposal based
+on a Channel concept:
+
+```python
+class Transport:
+  def __init__(self, sender: ray.ActorHandle, receiver: ray.ActorHandle):
+    pass
+  def send(self, send_meta: BufferMetadata):
+    """Ideally async. Called by the sender"""
+    pass
+  def recv(self, recv_meta: BufferMetadata) -> Buffer:
+    """Ideally async. Called by the receiver."""
+    pass
+```
+
+When an application-defined transport is used, we will use the default
+transport to synchronize between the sender and receiver, but we only
+send the BufferMetadata in along the default transport. This will act as
+a signal to begin the actual \`Transport.send\` and \`Transport.recv\`
```

Comment:
Isn't this signal very high overhead (basically we will have default transport + custom transport overhead)? Is it necessary if the transport happens under the channel (as an output of each DAG node)?